### DISCUSSION TITLE
envkey-php: ENVKEY invalid. Couldn't load vars.

### DISCUSSION THREAD

**Discussion by nikalas** [envkey/envkey#92] [Q&A] [answered]
I'm trying to incorporate envkey-php into my WordPress theme by functions.php, however I'm getting the error
`Fatal error: Uncaught Exception: ENVKEY invalid. Couldn't load vars.`

Same thing if I try to load it from index.php or wp-config.php in the WordPress root

I have no problem loading other composer packages, like dotenv (not at the same time though) which is what we're trying to migrate from.

I'm running, under MAMP on mac:
Apache 2.4.46
PHP 8.1
Composer 2.6.8

The thing is, if I run `vendor/envkey/envkey-php/ext/envkey-source_2.4.2_darwin_arm64/envkey-source --json --mem-cache --client-name envkey-php --client-version 2.4.3`, which is the command ran by Fetcher.php I get the expected keys.

If I load envkey-php from composer in my localhost site (nothing else but phpinfo()), everything works as expected as well.

**Comment by danenania** [comment]
Hey @nikalas, thanks for providing details on this.

> The thing is, if I run vendor/envkey/envkey-php/ext/envkey-source_2.4.2_darwin_arm64/envkey-source --json --mem-cache --client-name envkey-php --client-version 2.4.3, which is the command ran by Fetcher.php I get the expected keys.

When you say you can run this successfully, are you running it from the shell or directly within your php app? 

Where are you setting your `ENVKEY` environment variable? Can you confirm that it's set correctly in your php app? It sounds like it may not be getting through to the `envkey-source` system call.

**Reply by nikalas** [reply]
When I run it successfully I do it manually from the shell. When I log the result from the system call in envkey-php (from php) it finds nothing, the system call just returns the same error and then crashes, so ENVKEY env variable is never set inside php.

I simply ran `envkey init` in the project root (same dir as the I run `composer install`). I also made a copy of .envkey as .env but this made no difference.

**Comment by danenania** [comment]
To test an idea on this, could you try adding the following prior to pulling in the `envkey-php` package:

```
// Check if ENVKEY is available in $_SERVER or $_ENV and set it as an environment variable
if (isset($_SERVER['ENVKEY'])) {
    putenv("ENVKEY={$_SERVER['ENVKEY']}");
} elseif (isset($_ENV['ENVKEY'])) {
    putenv("ENVKEY={$_ENV['ENVKEY']}");
}
```

**Reply by nikalas** [reply]
This worked. How and why? Should it not be enough to just have the .envkey in the same folder? Which by the way does nothing if I remove it.

**Reply by danenania** [reply]
Great! We will add this to envkey-php and get it released shortly.

The problem seems to be that while system environment variables are forwarded to system calls (like the one to `envkey-source`) in php, env vars set in `$_SERVER` and `$_ENV` are not forwarded. So it seems that your `ENVKEY` is being set in either `$_SERVER` or `$_ENV` for some reason. Calling `putenv` ensures that the `ENVKEY` does get passed along to the `envkey-source` call.

**Reply by danenania** [reply]
I'd guess it has something to do with how Wordpress handles environment variables.

I'm assuming this was in the context of a development environment and not on a server?

**Reply by nikalas** [reply]
Yes, this is local dev server. I will be moving forward with testing on a staging server as I confirm this is working. Just as an fyi the key seems to be available as both `ENVKEY` and `REDIRECT_ENVKEY` in both `$_SERVER` and `$_ENV`

**Reply by nikalas** [reply]
No, this does not seem correct. Tried it from another project now and the syscall in Fetcher.php once again returns `echo 'error: ENVKEY missing '; false`. 

I have been throwing keys a bit of everywhere trying to get it to work earlier today, so I must have missed cleaning it up somewhere.

Edit: I seem to have set the local key manually in apache `httpd.conf`, which is why it was available in in `$_SERVER` and `$_ENV`. Removing those lines gets me back to not working again.

**Comment by nikalas** [comment, marked as answer]
Seems the problem is running PHP under Apaches Virtual Host (With FastCGI?) which runs sandboxed. As you say the env's are not exposed to the php service because of this. Which I guess is actually in the ENVKEY documentation. Also nothing to do with WordPress per se.

So the solution here seems to be to, if I understand it correctly:
- Generate a local key for each local app.
- Put it in MAMP under "Additional parameters for <VirtualHost> directive" for each site as `SetEnv ENVKEY <local-key-here>`
- Insert your snippet before `require_once 'vendor/autoload.php'` (I still don't see why this is needed as I can `getenv('ENVKEY')` and get the local key earlier, but it does not work without it.)

Thank you for you help!

**Reply by danenania** [reply]
Fyi, that snippet has now been included in envkey-php 2.4.4, so you shouldn't need it in your code any longer after upgrading.
